datafusion/functions/src/regex/regexpcount.rs

```diff
@@ -0,0 +1,766 @@
+                    values.len(),
+                    start_array.len()
+                )));
+            }
+
+            let flags_array = flags_array.unwrap();
+            if values.len() != flags_array.len() {
+                return Err(ArrowError::ComputeError(format!(
+                    "flags_array must be the same length as values array; got {} and {}",
+                    values.len(),
+                    flags_array.len()
+                )));
+            }
+
+            let mut regex_cache = HashMap::new();
+            Ok(Arc::new(Int64Array::from_iter_values(
+                izip!(
+                    values.iter(),
+                    regex_array.iter(),
+                    start_array.iter(),
+                    flags_array.iter()
+                )
+                .map(|(value, regex, start, flags)| {
+                    let regex = match regex {
+                        None | Some("") => return Ok(0),
+                        Some(regex) => regex,
+                    };
+
+                    let pattern =
+                        compile_and_cache_regex(regex, flags, &mut regex_cache)?;
+                    count_matches(value, &pattern, start)
+                })
+                .collect::<Result<Vec<i64>, ArrowError>>()?,
+            )))
+        }
+    }
+}
+
+fn compile_and_cache_regex(
+    regex: &str,
+    flags: Option<&str>,
+    regex_cache: &mut HashMap<String, Regex>,
+) -> Result<Regex, ArrowError> {
+    match regex_cache.entry(regex.to_string()) {
+        Entry::Vacant(entry) => {
+            let compiled = compile_regex(regex, flags)?;
+            entry.insert(compiled.clone());
+            Ok(compiled)
+        }
+        Entry::Occupied(entry) => Ok(entry.get().to_owned()),
+    }
+}
+
+fn compile_regex(regex: &str, flags: Option<&str>) -> Result<Regex, ArrowError> {
+    let pattern = match flags {
+        None | Some("") => regex.to_string(),
+        Some(flags) => {
+            if flags.contains("g") {
+                return Err(ArrowError::ComputeError(
+                    "regexp_count() does not support global flag".to_string(),
+                ));
+            }
+            format!("(?{}){}", flags, regex)
+        }
+    };
+
+    Regex::new(&pattern).map_err(|_| {
+        ArrowError::ComputeError(format!(
+            "Regular expression did not compile: {}",
+            pattern
+        ))
+    })
+}
+
+fn count_matches(
+    value: Option<&str>,
+    pattern: &Regex,
+    start: Option<i64>,
+) -> Result<i64, ArrowError> {
+    let value = match value {
+        None | Some("") => return Ok(0),
+        Some(value) => value,
+    };
+
+    if let Some(start) = start {
+        if start < 1 {
+            return Err(ArrowError::ComputeError(
+                "regexp_count() requires start to be 1 based".to_string(),
+            ));
+        }
+
+        let find_slice = value.chars().skip(start as usize - 1).collect::<String>();
```

Comment:
I wonder if it would be worth it to have a fast path if start is 1?  Untested suggestion:
```suggestion
        let find_slice = if start == 1 { start } else { value.chars().skip(start as usize - 1).collect::<String>()} ;
```